File: backend/app/schemas/planning.py
"""Pydantic schemas for the planning system API."""

from datetime import datetime
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


# -------------------------------------------------------------------------
# Request Schemas
# -------------------------------------------------------------------------


class PlanningSessionCreateRequest(BaseModel):
    """Request body for creating a new planning session."""

    project_id: str
    chat_id: Optional[str] = None
    title: str = Field(..., min_length=1, max_length=500)
    description: Optional[str] = None
    target_type: str = Field(default="sandbox", pattern="^(sandbox|ui_builder|both)$")
    success_criteria: List[str] = Field(default_factory=list)


class PlanningSessionUpdateRequest(BaseModel):
    """Request body for updating a planning session."""

    title: Optional[str] = Field(default=None, max_length=500)
    description: Optional[str] = None
    target_type: Optional[str] = Field(default=None, pattern="^(sandbox|ui_builder|both)$")
    status: Optional[str] = Field(
        default=None, pattern="^(draft|active|in_progress|completed|archived)$"
    )
    success_criteria: Optional[List[str]] = None


class PlanPhaseCreateRequest(BaseModel):
    """Request body for creating a new phase."""

    title: str = Field(..., min_length=1, max_length=500)
    description: str = ""
    inputs: List[str] = Field(default_factory=list)
    outputs: List[str] = Field(default_factory=list)
    implementation_plan: Dict[str, Any] = Field(default_factory=dict)
    verification_checks: List[Dict[str, Any]] = Field(default_factory=list)


class PlanPhaseUpdateRequest(BaseModel):
    """Request body for updating a phase."""

    title: Optional[str] = Field(default=None, max_length=500)
    description: Optional[str] = None
    inputs: Optional[List[str]] = None
    outputs: Optional[List[str]] = None
    implementation_plan: Optional[Dict[str, Any]] = None
    verification_checks: Optional[List[Dict[str, Any]]] = None
    status: Optional[str] = Field(
        default=None,
        pattern="^(pending|in_progress|verifying|completed|failed)$",
    )


class PlanTaskCreateRequest(BaseModel):
    """Request body for creating a new task."""

    title: str = Field(..., min_length=1, max_length=500)
    description: str = ""
    task_type: str = Field(
        ...,
        pattern="^(file_create|file_modify|file_delete|ui_component|ui_layout|ui_style|run_command|install_package|verification)$",
    )
    task_data: Dict[str, Any] = Field(default_factory=dict)
    depends_on: List[str] = Field(default_factory=list)


class PlanTaskUpdateRequest(BaseModel):
    """Request body for updating a task."""

    title: Optional[str] = Field(default=None, max_length=500)
    description: Optional[str] = None
    task_data: Optional[Dict[str, Any]] = None
    depends_on: Optional[List[str]] = None
    status: Optional[str] = Field(
        default=None,
        pattern="^(pending|ready|in_progress|completed|failed)$",
    )


class UIBuilderImportRequest(BaseModel):
    """Request body for importing UI builder state into a plan."""

    ui_tree: List[Dict[str, Any]]


# -------------------------------------------------------------------------
# Response Schemas
# -------------------------------------------------------------------------


class PlanTaskResponse(BaseModel):
    """Response for a single task."""

    id: str
    phase_id: str
    title: str
    description: str
    task_order: int
    task_type: str
    task_data: Optional[Dict[str, Any]] = None
    depends_on: Optional[List[str]] = None
    status: str
    result: Optional[Dict[str, Any]] = None
    automation_action_id: Optional[str] = None
    completed_at: Optional[str] = None
    created_at: str
    updated_at: str


class PlanPhaseResponse(BaseModel):
    """Response for a single phase."""

    id: str
    session_id: str
    title: str
    description: str
    phase_order: int
    inputs: Optional[List[str]] = None
    outputs: Optional[List[str]] = None
    implementation_plan: Optional[Dict[str, Any]] = None
    verification_checks: Optional[List[Dict[str, Any]]] = None
    status: str
    user_approved: bool
    verification_result: Optional[Dict[str, Any]] = None
    started_at: Optional[str] = None
    completed_at: Optional[str] = None
    tasks: List[PlanTaskResponse] = Field(default_factory=list)
    created_at: str
    updated_at: str


class PlanningSessionResponse(BaseModel):
    """Response for a planning session (list view)."""

    id: str
    project_id: str
    chat_id: Optional[str] = None
    user_id: str
    title: str
    description: Optional[str] = None
    target_type: str
    status: str
    current_phase_id: Optional[str] = None
    success_criteria: Optional[List[str]] = None
    phase_count: int = 0
    completed_phase_count: int = 0
    created_at: str
    updated_at: str


class PlanningSessionDetailResponse(PlanningSessionResponse):
    """Response for a planning session with full phase/task details."""

    phases: List[PlanPhaseResponse] = Field(default_factory=list)
    ui_builder_state: Optional[Dict[str, Any]] = None


class PlanProgressResponse(BaseModel):
    """Summary of plan execution progress."""

    session_id: str
    status: str
    total_phases: int
    completed_phases: int
    current_phase_title: Optional[str] = None
    total_tasks: int
    completed_tasks: int
    progress_percentage: float

File: frontend/packages/api/types/planning.ts
/** Types for the Traycer-style planning system. */

export interface PlanTask {
  id: string;
  phase_id: string;
  title: string;
  description: string;
  task_order: number;
  task_type: string;
  task_data: Record<string, unknown> | null;
  depends_on: string[] | null;
  status: "pending" | "ready" | "in_progress" | "completed" | "failed";
  result: Record<string, unknown> | null;
  automation_action_id: string | null;
  completed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface PlanPhase {
  id: string;
  session_id: string;
  title: string;
  description: string;
  phase_order: number;
  inputs: string[] | null;
  outputs: string[] | null;
  implementation_plan: Record<string, unknown> | null;
  verification_checks: Array<{ type: string; criteria: string }> | null;
  status: "pending" | "in_progress" | "verifying" | "completed" | "failed";
  user_approved: boolean;
  verification_result: Record<string, unknown> | null;
  started_at: string | null;
  completed_at: string | null;
  tasks: PlanTask[];
  created_at: string;
  updated_at: string;
}

export interface PlanningSession {
  id: string;
  project_id: string;
  chat_id: string | null;
  user_id: string;
  title: string;
  description: string | null;
  target_type: "sandbox" | "ui_builder" | "both";
  status: "draft" | "active" | "in_progress" | "completed" | "archived";
  current_phase_id: string | null;
  success_criteria: string[] | null;
  phase_count: number;
  completed_phase_count: number;
  created_at: string;
  updated_at: string;
}

export interface PlanningSessionDetail extends PlanningSession {
  phases: PlanPhase[];
  ui_builder_state: Record<string, unknown> | null;
}

export interface PlanProgress {
  session_id: string;
  status: string;
  total_phases: number;
  completed_phases: number;
  current_phase_title: string | null;
  total_tasks: number;
  completed_tasks: number;
  progress_percentage: number;
}

export interface PlanningSessionCreateRequest {
  project_id: string;
  chat_id?: string;
  title: string;
  description?: string;
  target_type?: "sandbox" | "ui_builder" | "both";
  success_criteria?: string[];
}

export interface PlanPhaseCreateRequest {
  title: string;
  description?: string;
  inputs?: string[];
  outputs?: string[];
  implementation_plan?: Record<string, unknown>;
  verification_checks?: Array<{ type: string; criteria: string }>;
}

export interface PlanTaskCreateRequest {
  title: string;
  description?: string;
  task_type: string;
  task_data?: Record<string, unknown>;
  depends_on?: string[];
}
